src/klein/_request.py

```diff
@@ -0,0 +1,73 @@
+# -*- test-case-name: klein.test.test_request -*-
+# Copyright (c) 2017-2018. See LICENSE for details.
+
+"""
+HTTP request API.
+"""
+
+from typing import Text, Union
+
+from attr import Factory, attrib, attrs
+from attr.validators import instance_of, provides
+
+from hyperlink import URL
+
+from tubes.itube import IFount
+
+from twisted.internet.defer import Deferred
+
+from zope.interface import Attribute, implementer
+
+from ._headers import IHTTPHeaders
+from ._message import (
+    IHTTPMessage, MessageState, bodyAsBytes, bodyAsFount, validateBody
+)
+
+# Silence linter
+Deferred, IFount, IHTTPHeaders, Text, Union
+
+
+__all__ = ()
+
+
+
+# Interfaces
+
+class IHTTPRequest(IHTTPMessage):
+    """
+    HTTP request.
+    """
+
+    method = Attribute("Request method.")  # type: Text
+    uri    = Attribute("Request URI.")     # type: URL
+
+
+
+# Implementation
+
+@implementer(IHTTPRequest)
+@attrs(frozen=True)
+class FrozenHTTPRequest(object):
+    """
+    Immutable HTTP request.
+    """
+
+    method  = attrib(validator=instance_of(Text))       # type: Text
+    uri     = attrib(validator=instance_of(URL))        # type: URL
+    headers = attrib(validator=provides(IHTTPHeaders))  # type: IHTTPHeaders
+
+    _body = attrib(validator=validateBody)  # type: Union[bytes, IFount]
+
+    _state = attrib(
+        default=Factory(MessageState), init=False
```

Comment:
So not really immutable

Reply:
Well, it's immutable insofar as there's no API that mutates it, though the presence of `fountExhausted` is arguably a mutation, in that the fount can only be accessed once, so accessing it causes a mutation.  I'll think on that some more.